File: GoTracer/src/Types/Vector.go
package types

import "math"

type Vector struct {
	X, Y, Z float64
}

func (v Vector) Dot(v1 Vector) float64 {
	return (v.X * v1.X) + (v.Y * v1.Y) + (v.Z * v1.Z)
}

func (v Vector) Length() float64 {
	return math.Sqrt(v.Dot(v))
}

func (v Vector) Cross(v1 Vector) Vector {
	return Vector{X: v.Y*v1.Z - v.Z*v1.Y,
		Y: v.Z*v1.X - v.X*v1.Z,
		Z: v.X*v1.Y - v.Y*v1.X}
}

func (v Vector) Add(v1 Vector) Vector {
	return Vector{X: v.X + v1.X, Y: v.Y + v1.Y, Z: v.Z + v1.Z}
}

func (v Vector) Subtract(v1 Vector) Vector {
	return Vector{X: v.X - v1.X, Y: v.Y - v1.Y, Z: v.Z - v1.Z}
}

func (v Vector) Scale(s float64) Vector {
	return Vector{X: v.X * s, Y: v.Y * s, Z: v.Z * s}
}

func (v Vector) Normalize() Vector {
	mag := v.Length()
	if mag == 0.0 {
		return v.Scale(math.Inf(1))
	} else {
		return v.Scale(1.0 / mag)
	}
}

File: SwiftTracer/SwiftTracer/Vector.swift
import Foundation

struct Vector {
    let x,y,z: Double;
    init(x:Double, y:Double, z:Double) {
        self.x = x;
        self.y = y;
        self.z = z;
    }

    func Length() -> Double {
        return sqrt(self.Dot(self))
    }
    
    func Dot(v:Vector) -> Double {
        return (self.x * v.x) + (self.y * v.y) + (self.z * v.z)
    }
    
    func Cross(v:Vector) -> Vector {
        return Vector(x: (self.y*v.z) - (self.z*v.y),
                      y: (self.z*v.x) - (self.x*v.z),
                      z: (self.x*v.y) - (self.y*v.x))
    }
    
    func Add(v:Vector) -> Vector {
        return Vector(x:self.x + v.x, y:self.y + v.y, z:self.z + v.z)
    }
    
    func Subtract(v:Vector) -> Vector {
        return Vector(x:self.x - v.x, y:self.y - v.y, z:self.z - v.z)
    }
    
    func Scale(s:Double) -> Vector {
        return Vector(x:self.x * s, y:self.y * s, z:self.z * s)
    }
    
    func Normalize() -> Vector {
        let mag = self.Length()
        if mag == 0 {
            return self.Scale(Double.infinity)
        }
        else {
            return self.Scale(1.0 / mag)
        }
    }
}